akka-distributed-data/src/main/scala/akka/cluster/ddata/DurableStore.scala

```diff
@@ -0,0 +1,257 @@
+import org.mapdb.DataOutput2
+import org.mapdb.HTreeMap
+import org.mapdb.Serializer
+
+/**
+ * An actor implementing the durable store for the Distributed Data `Replicator`
+ * have to implement the protocol with the messages defined here.
+ *
+ * At startup the `Replicator` creates the durable store actor and sends the
+ * `Load` message to it. It must then reply with 0 or more `LoadData` messages
+ * followed by one `LoadCompleted` message to the `sender` (the `Replicator`).
+ *
+ * If the `Load` fails it can throw `LoadFailed` and the `Replicator` supervisor
+ * will stop itself and the durable store.
+ *
+ * When the `Replicator` needs to store a value it sends a `Store` message
+ * to the durable store actor, which must then reply with the `successMsg` or
+ * `failureMsg` to the `replyTo`.
+ */
+object DurableStore {
+  final case class Store(key: String, data: ReplicatedData, reply: Option[StoreReply])
+  final case class StoreReply(successMsg: Any, failureMsg: Any, replyTo: ActorRef)
+
+  case object Load
+  final case class LoadData(data: Map[String, ReplicatedData])
+  case object LoadCompleted
+  class LoadFailed(message: String, cause: Throwable) extends RuntimeException(message) {
+    def this(message: String) = this(message, null)
+  }
+}
+
+object MapDbDurableStore {
+  def props(config: Config): Props =
+    Props(new MapDbDurableStore(config))
+
+  private case object Commit extends DeadLetterSuppression
+}
+
+final class MapDbDurableStore(config: Config) extends Actor with ActorLogging {
+  import DurableStore._
+  import MapDbDurableStore.Commit
+
+  val serializer = SerializationExtension(context.system).serializerFor(classOf[DataEnvelope])
+  val commitInterval = config.getString("mapdb.commit-interval").toLowerCase match {
+    case "off" ⇒ Duration.Zero
+    case _     ⇒ config.getDuration("mapdb.commit-interval", MILLISECONDS).millis
+  }
+
+  val db = {
+    val file = config.getString("mapdb.file") match {
+      case path if path.endsWith(".ddata") ⇒
+        new File(path)
+      case dir ⇒
+        new File(dir, s"${context.system.name}-${self.path.parent.name}-${Cluster(context.system).selfAddress.port.get}.ddata")
+    }
+    file.getParentFile.mkdirs()
+
+    val builder = DBMaker.fileDB(file)
+      .fileMmapEnableIfSupported()
+      .fileMmapPreclearDisable()
+      .allocateStartSize(config.getBytes("mapdb.allocate-start-size"))
+      .allocateIncrement(config.getBytes("mapdb.allocate-increment"))
+      .concurrencyDisable() // only access from this actor
+      .closeOnJvmShutdown()
+
+    if (config.getBoolean("mapdb.write-ahead-log"))
+      builder.transactionEnable() // WAL for crash safety
+
+    builder.make()
+  }
+  val store: HTreeMap[String, ReplicatedData] =
+    db.hashMap("store")
+      .keySerializer(Serializer.STRING)
+      .valueSerializer(new MapdbReplicatedDataSerializer(context.system))
+      .createOrOpen()
+
+  val uncommitted = new java.util.HashMap[String, ReplicatedData]
```

Comment:
this changes ordering for pending writes (and coalesces multiple writes to same location). I guess this is safe, but can you confirm?